### PR TITLE
Add FATES land use change module to pass LUH2 data to FATES

### PR DESCRIPTION
**Pull request by glemieux**
This pull request adds a new module, the structure of which is adopted from the dyn_subgrid code, 
to enable e3sm to ingest minimally processed luh2 data to be passed to FATES.  
A new namelist variable is introduced `use_fates_luh`, to engage the module functionality, 
which is off by default.  The luh2 dataset to be ingested is a minimally processed concatenation 
of the luh2 state, transition and management datasets.

These changes also include a minor bug fix discovered in the process of developing the code, 
in which if the number of FATES patches being defined by the user in the fates parameter file is greater than the `max_patch_per_col` elm variable, would result in a BalanceCheck error.  

To be coordinated with https://github.com/NGEET/fates/pull/1040

[nonBFB] for FATES

### CODE REVIEW

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (bishtgautam) [inline components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:135]: Please indent this code block.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:1023]: Please delete the commented out code and indent the code.

**Comment by bishtgautam** [comment]
@glemieux Can we add a new test for this PR?

**Comment by glemieux** [comment]
Status update: there is a forthcoming update to the fates luh2 data pipeline https://github.com/NGEET/fates/pull/1032 that will result in an update to the default fates landuse data set.  As such I've added "WIP" to this PR.

**Comment by rljacob** [comment]
status:  probably another month waiting on FATES refactor.

**Comment by rljacob** [comment]
Need #5849 and also still WIP.

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (glemieux) [reply components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90]: completed

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: done

**Comment by rljacob** [comment]
#5849 was merged.  Status now?

**Comment by rljacob** [comment]
@glemieux is this ready?

**Comment by glemieux** [comment]
> @glemieux is this ready?

This is still WIP and waiting on https://github.com/NGEET/fates/pull/1040  updates.  I'm actively working on this.

**Comment by rljacob** [comment]
update: still waiting on fates PR.  Another couple of weeks.

**Comment by rljacob** [comment]
update:  still WIP.  2 others will need to be merged before this.

**Comment by rljacob** [comment]
@glemieux can you list in a comment the PRs this is waiting on?

**Comment by glemieux** [comment]
@rljacob it needs to be staged after #6018 and #6027, both which need review and are undergoing some testing.  Ultimately, it's also waiting on https://github.com/NGEET/fates/pull/1040, but this should be integrated within the next week.  

This PR needs to have some updates brought in, but is close to not being "WIP".

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90
Comment (rgknox) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:757]: great, glad this fix is in

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (rgknox) [inline components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:118]: repeated text, ok to remove?

**Review by rgknox — COMMENTED** [review]
@glemieux and I are discussing some cosmetic things and checking on a setting in the namelist build perl code, but otherwise I think this PR looks good.  @glemieux plans to add a test for this as well

**Comment by rgknox** [comment]
FYI: The next FATES API (33) is fairly simple, but I will hold on making that until this branch is re-based, since I have to build it on top of this one.

**Comment by glemieux** [comment]
- [ ] add landuse testmod to `fates` test suite
- [ ] remove `use_cn` return check from landusemod
- [ ] add `fluh_timeseries` to `setup_cmdl_fates_mode`

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (glemieux) [reply components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90]: Totally; Removed.

**Comment by glemieux** [comment]
@peterdschwartz @rgknox I've taken care of the suggested changes and  rebased this onto `master`.  I'm going to start regression testing on this today and I've removed the '[WIP]'.

**Comment by glemieux** [comment]
@peterdschwartz regression testing on perlmutter with the `e3sm_land_developer` suite is complete and all expected non-fates tests are B4B.  The `qian` smoke test from #6192 is failing, but that's because I hadn't rebased since last week.  I've got a local rebased branch with #6219 included to make sure its not failing for a different reason.  All fates test pass with expected DIFFs in non-landuse run modes since the tag update also includes a couple fates scientific bug fixes since `sci.1.68.2_api.31.0.0`.

**Comment by glemieux** [comment]
Just a quick update to note that rebasing against the latest `master` resulted in `SMS.r05_r05.I1850ELMCN.pm-cpu_....elm-qian_1948` passing.

**Comment by peterdschwartz** [comment]
Tested on chrysalis and everything came back as expected.  This can go in once it's got one approval

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (bishtgautam) [inline components/elm/bld/namelist_files/namelist_defaults.xml:570]: 1. This file is not on the E3SM input data server.
2. FATES's SCI/API version number will continue to get updated. This would mean we would have many `fates-sci*` directories. How we have a single dir (e.g. `lnd/clm2/surfdata_map/fates`) and the metadata of the LUH2 file can include the information regarding the SCI/API version. In the F90 code, you can also include a sanity check regarding the SCI/API version.

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (bishtgautam) [inline components/elm/bld/namelist_files/namelist_defaults.xml:130]: This file is not available on https://web.lcrc.anl.gov/public/e3sm/inputdata/lnd/clm2/paramdata/

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (bishtgautam) [inline components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:34]: How about adding the long name for these in the comment?

```
  character(len=5), public, parameter  :: landuse_state_varnames(num_landuse_state_vars) = &
                    [character(len=5)  :: 'primf', & ! Primary forest 
                                          'primn', & ! Primary something???
```

**Review by bishtgautam — CHANGES_REQUESTED** [review]
I have a few minor comments. However, a major comment is: Shouldn't there be a new compset (say I20TRELMFATES) in which the transient LU data is used by FATES?

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (glemieux) [reply components/elm/bld/namelist_files/namelist_defaults.xml:570]: 1. @bishtgautam this file is fairly large.  I submitted a request about 2 weeks ago to get access to the ANL servers using the procedure laid out on confluence so that I could load this and future updates directly.  I haven't seen movement on that request.  In the meantime, the data is on `perlmutter`; is this something you could upload for me via globus?
2.  That should work.  The file structure here was the recommended by ctsm folks, although I can't remember the rationale currently.  I'll definitely add the fates sci/api version to the data file meta data.  That said, the meta data check isn't really needed for usage here.

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (glemieux) [reply components/elm/bld/namelist_files/namelist_defaults.xml:130]: Thanks @bishtgautam I'll email you the file.

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (glemieux) [reply components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:34]: That's a good idea.  I could also point developers to the [data readme](https://luh.umd.edu/LUH2/LUH2_v2h_README.pdf) location online.  How do we feel about having urls in the code comments?

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (peterdschwartz) [reply components/elm/bld/namelist_files/namelist_defaults.xml:130]: The file is on the other server though.  https://svn-ccsm-inputdata.cgd.ucar.edu/trunk/inputdata/lnd/clm2/paramdata/

Comment (bishtgautam) [reply components/elm/bld/namelist_files/namelist_defaults.xml:130]: Oh, FATES also contributes code to CTSM, so it is on NCAR's server. In that case, we don't need to put on the E3SM server.

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (bishtgautam) [reply components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:34]: I believe the URLs are discouraged because they would inevitably get broken in the future.

File: components/elm/bld/namelist_files/namelist_defaults.xml
Comment (glemieux) [reply components/elm/bld/namelist_files/namelist_defaults.xml:130]: Oh nice, I didn't realize that!  Thanks @peterdschwartz

File: components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90
Comment (glemieux) [reply components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90]: Updated via f09f5e3b8bf54ce5eb13e7597946cfbcf4a7f39e

**Review by rgknox — APPROVED** [review]

**Comment by glemieux** [comment]
> However, a major comment is: Shouldn't there be a new compset (say I20TRELMFATES) in which the transient LU data is used by FATES?

Eventually we should have a compset for this, but with this PR I don't think its necessary as we're going to be bringing in a fates landuse v2 change in the next month or so.  I think with the v2 update we will want to facilitate easier use of the option through the compset.  

On a related note and by comparison, I think its definitely worth creating a fates SP mode compset with a future PR since that run mode is more mature.

**Review by bishtgautam — APPROVED** [review]

**Comment by peterdschwartz** [comment]
Still waiting for the repo to be open.

**Pull request merged**
